compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/insert_aie_workgroup.mlir

```diff
@@ -0,0 +1,98 @@
+    %c64 = arith.constant 64 : index
+    %c0_i32 = arith.constant 0 : i32
+    %c0 = arith.constant 0 : index
+    %alloc = memref.alloc() : memref<4x8x8x8xi32, 2>
+    %alloc_0 = memref.alloc() : memref<8x8x4x8xi32, 2>
+    %alloc_1 = memref.alloc() : memref<4x8x4x8xi32, 2>
+    %alloc_2 = memref.alloc() : memref<64x32xi32, 1>
+    %alloc_3 = memref.alloc() : memref<32x64xi32, 1>
+    %alloc_4 = memref.alloc() : memref<1024x64xi32>
+    %alloc_5 = memref.alloc() : memref<32x1024xi32>
+    %alloc_6 = memref.alloc() : memref<32x32xi32, 1>
+    %alloc_7 = memref.alloc() : memref<32x64xi32>
+    %0 = amdaie.logicalobjectfifo.from_memref %alloc_4, {} : memref<1024x64xi32> -> !amdaie.logicalobjectfifo<memref<1024x64xi32>>
+    %1 = amdaie.logicalobjectfifo.from_memref %alloc_5, {} : memref<32x1024xi32> -> !amdaie.logicalobjectfifo<memref<32x1024xi32>>
+    %2 = amdaie.logicalobjectfifo.from_memref %alloc_2, {} : memref<64x32xi32, 1> -> !amdaie.logicalobjectfifo<memref<64x32xi32, 1>>
+    %3 = amdaie.logicalobjectfifo.from_memref %alloc_3, {} : memref<32x64xi32, 1> -> !amdaie.logicalobjectfifo<memref<32x64xi32, 1>>
+    %4 = amdaie.logicalobjectfifo.from_memref %alloc, {} : memref<4x8x8x8xi32, 2> -> !amdaie.logicalobjectfifo<memref<4x8x8x8xi32, 2>>
+    %5 = amdaie.logicalobjectfifo.from_memref %alloc_0, {} : memref<8x8x4x8xi32, 2> -> !amdaie.logicalobjectfifo<memref<8x8x4x8xi32, 2>>
+    %6 = amdaie.logicalobjectfifo.from_memref %alloc_1, {} : memref<4x8x4x8xi32, 2> -> !amdaie.logicalobjectfifo<memref<4x8x4x8xi32, 2>>
+    %7 = amdaie.logicalobjectfifo.from_memref %alloc_6, {} : memref<32x32xi32, 1> -> !amdaie.logicalobjectfifo<memref<32x32xi32, 1>>
+    %8 = amdaie.logicalobjectfifo.from_memref %alloc_7, {} : memref<32x64xi32> -> !amdaie.logicalobjectfifo<memref<32x64xi32>>
+    scf.forall (%arg0, %arg1) in (1, 1) {
+      scf.forall (%arg2, %arg3) in (1, 2) {
+        %9 = affine.apply #map(%arg2)
+        %10 = affine.apply #map(%arg3)
+        %11 = amdaie.dma_cpy_nd(%3[] [] [], %1[%9, %c0] [%c32, %c64] [%c1024, %c1]) : (!amdaie.logicalobjectfifo<memref<32x64xi32, 1>>, !amdaie.logicalobjectfifo<memref<32x1024xi32>>)
+        %12 = amdaie.dma_cpy_nd(%2[] [] [], %0[%c0, %10] [%c64, %c32] [%c64, %c1]) : (!amdaie.logicalobjectfifo<memref<64x32xi32, 1>>, !amdaie.logicalobjectfifo<memref<1024x64xi32>>)
+        linalg.fill ins(%c0_i32 : i32) outs(%alloc_1 : memref<4x8x4x8xi32, 2>)
+        %13 = amdaie.dma_cpy_nd(%5[%c0, %c0, %c0, %c0] [%c8, %c8, %c4, %c8] [%c256, %c32, %c8, %c1], %3[%c0, %c0, %c0, %c0] [%c8, %c8, %c4, %c8] [%c8, %c256, %c64, %c1]) : (!amdaie.logicalobjectfifo<memref<8x8x4x8xi32, 2>>, !amdaie.logicalobjectfifo<memref<32x64xi32, 1>>)
+        %14 = amdaie.dma_cpy_nd(%4[%c0, %c0, %c0, %c0] [%c4, %c8, %c8, %c8] [%c512, %c64, %c8, %c1], %2[%c0, %c0, %c0, %c0] [%c4, %c8, %c8, %c8] [%c8, %c256, %c32, %c1]) : (!amdaie.logicalobjectfifo<memref<4x8x8x8xi32, 2>>, !amdaie.logicalobjectfifo<memref<64x32xi32, 1>>)
```

Comment:
Does it require all the `amdaie.dma_cpy_nd` within the inner scf.forall to be able to include in the workgroup? In real application, how do you make all these `amdaie.dma_cpy_nd` all under the inner scf.forall?  I think the first two should be outside the inner scf.forall.

Reply:
> I think the first two should be outside the inner scf.forall.

Yes, they mostly should be and if they end up outside the workgroup, they can be fused into the workgroup later, using:
https://github.com/nod-ai/iree-amd-aie/compare/main...jornt_peeled_matmul_e2e#diff-61d9a09130faccae2d9f693e42252c7fb1b1780b26b88f9502ff59e29e0a07e7R20.